### PR TITLE
CLN: de-duplicate generate_range

### PR DESCRIPTION
**Pull request by jbrockmendel**
Split off from #23140.

### CODE REVIEW

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -332,6 +332,11 @@ def _validate_frequency(cls, index, freq, **kwargs):
             # Frequency validation is not meaningful for Period Array/Index
             return None
 
+        # DatetimeArray may pass `ambiguous`, nothing else will be accepted
+        #  by cls._generate_range below
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: I don't really like this, can we just directly pass ``ambiguous`` as required?

Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: We could, but it is much more verbose (BTW the letter "v" on my keyboard is super-sticky, so we really need to stop talking about erbosity).

This would end up looking something like:

```
if is_timedelta64_dtype(index):
    assert ambiguous is None
    on_freq = cls._generate_range(start=index[0], end=None,
                                                        periods=len(index), freq=freq)
else:
    on_freq = cls._generate_range(start=index[0], end=None,
                                                        periods=len(index), freq=freq, ambiguous=ambiguous)
```

Or do you have something else in mind?  I like this less than the way the PR currently does it, but by a small enough margin that I'm fine changing it to get this done.

Comment (jreback) [reply pandas/core/arrays/datetimelike.py]: no what I would do is add this to DatetimeIndex. It doesn't totally de-duplicate but is more correct and less hacky here

```
@classmethod
    def _generate_range(cls, start, end, periods, name=None, freq=None,
                        tz=None, normalize=False, ambiguous='raise',
                        closed=None):
        return super(DatetimeIndex, cls)._generate_range(
            start, end, periods, freq,
            tz=tz, normalize=normalize, ambiguous=ambiguous, closed=closed)
```

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
> no what I would do is add this to DatetimeIndex. [...]

Um, I don't see how that is really relevant to the `_validate_frequency` usage.  `DatetimeArray._generate_range` already accepts `ambiguous`; it is a matter of passing that arg in `_validate_frequency`.

**Comment by jbrockmendel** [comment]
I mean, I could just take out the new assertion in `_validate_frequency` and the method wouldn't be changed at all by this PR.

**Comment by jreback** [comment]
> Um, I don't see how that is really relevant to the _validate_frequency usage. DatetimeArray._generate_range already accepts ambiguous; it is a matter of passing that arg in _validate_frequency.

then let's just make this the default. This is a very kludgy way of doing that.

File: pandas/core/indexes/datetimelike.py [hunk 2]
```diff
@@ -703,6 +703,12 @@ def astype(self, dtype, copy=True):
             raise TypeError(msg.format(name=type(self).__name__, dtype=dtype))
         return super(DatetimeIndexOpsMixin, self).astype(dtype, copy=copy)
 
+    @Appender(DatetimeLikeArrayMixin._time_shift.__doc__)
+    def _time_shift(self, periods, freq=None):
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/datetimelike.py]: how is this related to the rest of the PR?

**Review by jorisvandenbossche — COMMENTED** [review]
Looks OK to me

File: pandas/core/indexes/datetimelike.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: This PR changes DatetimeArrayMixin._time_shift such that it no longer pins `self.name` for Index subclasses.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
looks fine. not super happy about having to set name all over the place, but this is an improvement over before. merging after CI is fixed.

**Comment by jreback** [comment]
thanks!
